packages/core/useStorageAsync/index.ts

```diff
@@ -95,31 +96,47 @@ export function useStorageAsync<T extends(string | number | boolean | object | n
     }
   }
 
-  read()
+  read().then()
 
-  if (window && listenToStorageChanges)
-    useEventListener(window, 'storage', e => Promise.resolve().then(() => read(e)))
+  async function cb() {
+    try {
+      if (data.value == null)
+        await storage!.removeItem(key)
+      else
+        await storage!.setItem(key, await serializer.write(data.value))
+    }
+    catch (e) {
+      onError(e)
+    }
+  }
+
+  let watchStop: WatchStopHandle = () => {}
 
   if (storage) {
-    watchWithFilter(
-      data,
-      async () => {
-        try {
-          if (data.value == null)
-            await storage!.removeItem(key)
-          else
-            await storage!.setItem(key, await serializer.write(data.value))
-        }
-        catch (e) {
-          onError(e)
-        }
-      },
-      {
-        flush,
-        deep,
-        eventFilter,
-      },
-    )
+    watchStop = watchWithFilter(data, cb, {
+      flush,
+      deep,
+      eventFilter,
+    })
+  }
+
+  // Block time-consuming reads
+  const queue: TaskQueue = new TaskQueue(() => {
+    watchStop()
+    watchStop = watchWithFilter(data, cb, {
+      flush,
+      deep,
+      eventFilter,
+    })
+  })
+
+  if (window && listenToStorageChanges) {
+    useEventListener(window, 'storage', (e) => {
+      watchStop()
+      queue.enqueue(async () => {
+        await read(e)
+      })
+    })
```

Comment:
From what I gathered this could just be simplified like this
```suggestion
    useEventListener(window, 'storage', async (e) => {
      watchStop()
      await read(e)
      watchStop = watchWithFilter(data, cb, {
        flush,
        deep,
        eventFilter,
      })
    })
```

Reply:
Sorry to reply to you now, you are correct, I was trying to reduce the side effects of "setItem" at the time.